cpp/test/sg/rf_treelite_test.cu

```diff
@@ -0,0 +1,397 @@
+    // https://github.com/dmlc/treelite/blob/528d883f8f39eb5dd633e929b95915b63e210b39/python/treelite/contrib/__init__.py.
+    std::string obj_cmd = "gcc -c -O3 -o " + dir_name + "/main.o " + dir_name +
+                          "/main.c -fPIC "
+                          "-std=c99 -lm";
+
+    std::string lib_cmd = "gcc -shared -O3 -o " + dir_name +
+                          "/treelite_model.so " + dir_name +
+                          "/main.o -std=c99 -lm";
+
+    ASSERT(system(obj_cmd.c_str()) == 0, "Call %s fails.", obj_cmd.c_str());
+    ASSERT(system(lib_cmd.c_str()) == 0, "Call %s fails.", lib_cmd.c_str());
+
+    PredictorHandle predictor;
+    std::string lib_path = dir_name + "/treelite_model.so";
+
+    // -1 means use maximum possible worker threads.
+    int worker_thread = -1;
+    TREELITE_CHECK(
+      TreelitePredictorLoad(lib_path.c_str(), worker_thread, &predictor));
+
+    DenseBatchHandle dense_batch;
+    // Current RF dosen't seem to support missing value, put NaN to be safe.
+    float missing_value = std::numeric_limits<double>::quiet_NaN();
+    TREELITE_CHECK(TreeliteAssembleDenseBatch(
+      inference_data_h.data(), missing_value, params.n_inference_rows,
+      params.n_cols, &dense_batch));
+
+    // Use dense batch so batch_sparse is 0.
+    // pred_margin = true means to produce raw margins rather than transformed probability.
+    int batch_sparse = 0;
+    bool pred_margin = false;
+    // Allocate larger array for treelite predicted label with using multi-class classification to avoid seg faults.
+    // Altough later we only use first params.n_inference_rows elements.
+    size_t treelite_predicted_labels_size;
+
+    TREELITE_CHECK(TreelitePredictorPredictBatch(
+      predictor, dense_batch, batch_sparse, verbose, pred_margin,
+      treelite_predicted_labels.data(), &treelite_predicted_labels_size));
+
+    TREELITE_CHECK(TreeliteDeleteDenseBatch(dense_batch));
+    TREELITE_CHECK(TreelitePredictorFree(predictor));
+    TREELITE_CHECK(TreeliteFreeModel(model));
+  }
+
+  void getResultAndCheck() {
+    // Predict and compare against known labels
+    RF_metrics tmp =
+      score(handle, forest, inference_data_d, labels_d, params.n_inference_rows,
+            params.n_cols, predicted_labels_d, false);
+    CUDA_CHECK(cudaStreamSynchronize(stream));
+
+    if (is_classification) {
+      predicted_labels_h.resize(params.n_inference_rows);
+      CUDA_CHECK(cudaMemcpy(predicted_labels_h.data(), predicted_labels_d,
+                            sizeof(L) * params.n_inference_rows,
+                            cudaMemcpyDeviceToHost));
+    } else {
+      // In case of regression, predicted_labels_d alreay has the final result.
+      ref_predicted_labels.resize(params.n_inference_rows);
+      CUDA_CHECK(cudaMemcpy(ref_predicted_labels.data(), predicted_labels_d,
+                            sizeof(L) * params.n_inference_rows,
+                            cudaMemcpyDeviceToHost));
+    }
```

Comment:
Could you use `updateHost()` instead of `cudaMemcpy()`? You'll need `cudaStreamSynchronize()` after that.